RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+#include <alpaka/alpaka.hpp>
+
+#include "FWCore/Utilities/interface/bit_cast.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/atomicMaxF.h"
+
+#include "DataFormats/ParticleFlowReco/interface/PFLayer.h"
+#include "RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.h"
+#include "RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterECLCC.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  using namespace cms::alpakatools;
+
+  using namespace reco::pfClustering;
+
+  static const int threadsPerBlockForClustering = std::is_same_v<Device, alpaka::DevCpu> ? 32 : 512;
+  static constexpr uint32_t blocksForExoticClusters = 4;
+
+  static constexpr uint32_t kHBHalf = 1296;
+  static constexpr uint32_t maxTopoInput = 2 * kHBHalf;
+
+  // Calculation of dR2 for Clustering
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE static float dR2(Position4 pos1, Position4 pos2) {
+    float mag1 = sqrtf(pos1.x * pos1.x + pos1.y * pos1.y + pos1.z * pos1.z);
+    float cosTheta1 = mag1 > 0.0 ? pos1.z / mag1 : 1.0;
+    float eta1 = 0.5 * logf((1.0 + cosTheta1) / (1.0 - cosTheta1));
```

Comment:
```suggestion
    float eta1 = 0.5f * logf((1.0f + cosTheta1) / (1.0f - cosTheta1));
```
to avoid doing the intermediate steps in double precision (same below).
Unless it was intended ?

Also, since you explicitly check for `mag1 > 0.0` above, should you check here for `cosTheta1` being `1.0` or `-1.0` and put a cut-off value for `eta1` ?
Or is it OK to let it go to +/-infinity ?

Reply:
This particular `dR2` calculation predates me, so I am not so sure if all the `double`/`float` intermediate steps are intentional. As for the potential of infinities, I think we had just been lucky to not encounter this case thus far. To my eyes it looks like it would be problematic if we encounter this. I will work on putting in a check.